src/lib.rs

```diff
@@ -35,42 +35,112 @@
         let data = s.as_bytes();
         let len = data.len();
 
+        // Obtain info about the panic dump region
+        let start_ptr = unsafe { &mut _panic_dump_start as *mut u8 };
+        let end_ptr   = unsafe { &mut _panic_dump_end as *mut u8 };
+        let max_len   = end_ptr as usize - start_ptr as usize;
+        let max_len_str = max_len - size_of::<usize>() - size_of::<usize>();
+
+        // If we have written the full length of the region, we can't write any
+        // more. This could happen with multiple writes with this implementation
+        if self.offset >= max_len_str {
+            return Ok(());
+        }
+
+        // We should write the size of the string, or the amount of space
+        // we have remaining, whichever is less
+        let str_len = min(max_len_str - self.offset, len);
+
         unsafe {
-            core::ptr::copy(
+            // Write the magic word for later detection
+            start_ptr
+                .cast::<usize>()
+                .write(0x0FACADE0);
+
+            // For now, skip writing the length...
+
+            // Write the string to RAM
+            core::ptr::copy_nonoverlapping(
                 data.as_ptr() as *mut u8,
-                (&mut __sbss as *mut u8).offset(self.offset as isize),
-                len,
-            )
+                start_ptr.offset(8).offset(self.offset as isize),
+                str_len,
+            );
+
+            // Increment the offset so later writes will be appended
+            self.offset += str_len;
+
+            // ... and now write the current offset (or total size) to the size location
+            start_ptr
+                .offset(4)
+                .cast::<usize>()
+                .write(self.offset);
         };
 
-        self.offset += len as u32;
         Ok(())
     }
 }
 
+/// Get the panic message from the last boot, if any
+pub fn get_panic_message() -> Option<&'static str> {
+    // Obtain panic region start and end from linker symbol _panic_dump_start and _panic_dump_end
+    extern "C" {
+        static mut _panic_dump_start: u8;
+        static mut _panic_dump_end: u8;
+    }
+
+    let start_ptr = unsafe { &mut _panic_dump_start as *mut u8 };
+
+    if 0x0FACADE0 != unsafe { core::ptr::read(start_ptr.cast::<usize>()) } {
+        return None;
+    }
+
+    // Obtain info about the panic dump region
+    let end_ptr   = unsafe { &mut _panic_dump_end as *mut u8 };
+    let max_len   = end_ptr as usize - start_ptr as usize;
+    let max_len_str = max_len - size_of::<usize>() - size_of::<usize>();
+
+    let len = unsafe { core::ptr::read(start_ptr.offset(4).cast::<usize>()) };
+
+    if len > max_len_str {
+        return None;
+    }
+
+    // TODO: This is prooooooooobably undefined behavior
+    let byte_slice = unsafe { core::slice::from_raw_parts(
+        start_ptr.offset(8),
+        len
+    )};
+
+    Some(unsafe { core::str::from_utf8_unchecked(byte_slice) })
```

Comment:
Note: This is unsound, as we could truncate on a non UTF-8 character bound, leaving a bad-utf8 string in the memory region. Is there an efficient way to avoid this?

Reply:
Check if it is utf8, else reduce by 1 and try again? Not elegant but should do the trick

Reply:
Honestly I'd just filter out anything not printable as ASCII. Why bother with UTF-8?

Reply:
If say the main point being that not all use English, if someone wanted to write a panic message in their own language we should allow it.

Reply:
On embedded? Sure, if that floats your boat... Doesn't mean we have to support it and all suffer from the code bloat...

Reply:
Then we can have both, with the utf8 being standard, and a feature that makes it ASCII only. That way we cater to inclusiveness while being able to optimize if we want, sounds like a win-win.